packages/tools/fluid-runner/src/exportFile.ts

```diff
@@ -0,0 +1,99 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Loader } from "@fluidframework/container-loader";
+import { createLocalOdspDocumentServiceFactory } from "@fluidframework/odsp-driver";
+import { ChildLogger, PerformanceEvent } from "@fluidframework/telemetry-utils";
+import * as fs from "fs";
+import FileLogger from "./logger/FileLogger";
+import { getArgsValidationError } from "./getArgsValidationError";
+import { IFluidFileConverter, isCodeLoaderBundle } from "./codeLoaderBundle";
+import { FakeUrlResolver } from "./fakeUrlResolver";
+import path from "path";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+
+export type IExportFileResponse = IExportFileResponseSuccess | IExportFileResponseFailure;
+
+interface IExportFileResponseSuccess {
+    success: true;
+}
+
+interface IExportFileResponseFailure {
+    success: false;
+    errorMessage: string;
+}
+
+export async function exportFile(
+    codeLoader: string,
+    inputFile: string,
+    outputFolder: string,
+    scenario: string,
+    telemetryFile: string,
+): Promise<IExportFileResponse> {
+    if (fs.existsSync(telemetryFile)) {
+        return { success: false, errorMessage: `Telemetry file already exists [${telemetryFile}]` };
+    }
+
+    const fileLogger = new FileLogger(telemetryFile, 50);
+
+    const logger = ChildLogger.create(fileLogger, "LocalSnapshotRunnerApp",
+        { all: { "Event_Time": () => Date.now() } });
+
+    try {
+        await PerformanceEvent.timedExecAsync(logger, { eventName: "ExportFile" }, async () => {
+            const codeLoaderBundle = require(codeLoader);
+            if (!isCodeLoaderBundle(codeLoaderBundle)) {
+                const message = "Code loader bundle is not of type CodeLoaderBundle";
+                logger.sendErrorEvent({
+                    eventName: "Client_ArgsValidationError",
+                    message,
+                });
+                return { success: false, errorMessage: message };
+            }
+
+            const argsValidationError = getArgsValidationError(inputFile, outputFolder, scenario);
+            if (argsValidationError) {
+                logger.sendErrorEvent({
+                    eventName: "Client_ArgsValidationError",
+                    message: argsValidationError,
+                });
+                return { success: false, errorMessage: argsValidationError };
+            }
+
+            // TODO: read file stream
+            const inputFileContent = fs.readFileSync(inputFile, { encoding: "utf-8" });
+
+            const results = await createContainerAndExecute(inputFileContent, logger, await codeLoaderBundle.fluidExport);
```

Comment:
@leeviana I was chatting with Teja earlier about this code. If I understand correctly, this is where office-web-host may fit in? Here in this example, "inputFileContent" seems to be a string that's a snapshot. And "codeLoaderBundle.fluidExport" seems to be a construct to help us to load the proper container. I wonder if you can shed some light on how office-web-host may fit here (set aside the problem of office-web-host is not published to a public feed so this program can't access it directly).

Reply:
Here is my naive understanding. There are couple layers to it, and you can integrate at any layer, in priority order (i.e. the lower on the list, the less supported / ideal is the flow):
1. We give you IContainer. Best, because we control how all the pieces are tied together (URI resolver, driver, etc.).
2. We given you ILoader. Worse, because you need to know what request to use to create container (we should assume that the system is setup in a way where it creates container for only one file instance, and that instance is hard coded via specific URI, asking for any other URI to be resolved will produce an error).
3. We give you URI resolver, driver factory, URI to use, file content (buffer), maybe loader options, maybe something else, and you are responsible for tying it all together. Obviously, there are a lot of inner details that are exposed here which will make it extremely hard to change anything if we need to. I.e. if we need to switch to different workflow (let's say use different driver instance), then it will be more expensive to do so across whole stack.

@leeviana - would love to hear what are your thoughts here.

Reply:
Yes, let me also copy over my code snippet/response from the Teams chat here. Here's my suggested integration given this code:

Reply:
const hostConfig = { logger }
return loadComponentInDiv({
    containerRequest: { url: "/fakeUrl/" },
    scope: fluidFileConverter.scope,
    urlResolver: new FakeUrlResolver(),
    documentServiceFactory: createLocalOdspDocumentServiceFactory(localOdspSnapshot),
    hostConfig
});

Reply:
There will be no div - that's Node.js environment. Quickly glancing at the code, it references `window` object - this code should not compile as is (if Teja does proper work to ensure DOM types are not available in the code that compiles and packaged for this solution).

That's the list favorable option overall (irrespective of above) - your code should not care about all these details, and we should ideally not expose them.

Reply:
Please ignore the div naming, that's old, note that no div is passed in or assumed in these inputs. Things that have to do with the window are similarly likely related to only sourceless iframe code that wouldn't be used here. We probably should look at the codeloader code though since I believe the script download mechanism it uses involves loading the code onto the window and reading it - though I'm not well versed in any codeloading mechanisms that don't involve this.

From the perspective of the details, these are details that would be in this code here (see the implementation of createContainerAndExecute). I'm not adding in any new parameters that are not already exposed in this code (and could be changed in this code)

Reply:
My proposal is simply to replace your "createContainerAndExecute" call with the OWH "createContainerAndExecute" call. It is fundamentally exactly the same except it brings with it things like codeloader logic for all MSFT containers/JS containers and any nuanced/weird logical things needed like precisely placed awaits to handle fluid race condition bugs (and the removal of those awaits when those bugs can be fixed! etc.). These are all things you would need to manually discover and add here and we'd end up with two unnecessary copies of the same code/logic.

Reply:
I do not want to expose FakeUrlResolver, createLocalOdspDocumentServiceFactory, and how file content is encoded (is it one buffer, or some data structure, or a set of files on disk) - all of that should be implementation details that are not leaked out, allowing us to change it (for example, move to FileDeltaStorageService if that makes sense). Another thing we can do - get rid of Loader altogether in this flow - it's Ok to create Container directly, removing one extra layer (fake driver factory) and hard-coded value (container URI). I really want to be able to do all these changes without any impact or slow down and the need to sync up.

I'd say it's Ok to provide functor like below to app specific function (reversing somewhat the flow), if it helps:
```ts
(scope, codeLoader) => { container: IContainer; logger: ITelemetryBaseLogger }
```

I think you refer (by awaits) to cases where old content is loaded, and you need to wait for latest state of document. This is not applicable here - we work with fully offline version of the document. If there are any bugs here, I'd prefer to learn about them and fix them.

Reply:
I'm not proposing that we expose it any more than it's exposed in this file. We are referring to FakeUrlResolver here in this code and my code suggestion would live in the same place in the same file, not a different layer. From the perspective of the code consuming this file's capabilities, none of these details have to be exposed.

I don't think we really have a need to inject any app specific functionality into the loader, especially if there's no need for waitForOpCatchUp classes of problems (which I think is one of the last lingering ones we're working to get rid of - so it sounds like that fully offline version will always be caught up to the latest state?) All I'm suggesting is to reuse the OWH code which does exactly "new Loader" but with any shared telemetry/bugfixes etc.

There was also an option that I was working with someone on our team for introducing an optional Loader override in OWH's HostConfig, but it looks like that PR was abandoned - this is something it could make sense to do though, but again, my suggestion is for all of this code to live in the same layer so it shouldn't matter too much.

At the point in time when you're writing all the code to return a container, that is exactly what OWH does so you'd be replicating any shared hosting code needed at that layer and have no need for OWH since you have your own copy.